feat: implement `parallel_mergesort` function

Matthew-ew opened the pull request.

Matthew-ew committed:
feat: implement `parallel_mergesort` function

Izak8 commented on mergesort.c:
```diff
@@ -56,8 +56,31 @@ void my_mergesort(int left, int right) {
+	/* base case: array has one element left */
+	if (argument->level == 0 || argument->right == 0) {
+		my_mergesort(argument->left, argument->right);
+	} else {
```
```suggestion
	if (argument->level == 0) {
		my_mergesort(argument->left, argument->right);
	}
	else {
	    if (argument->left >= argument->right) { return NULL; }
```
I think this is what is needed.

Izak8 commented on mergesort.c:
```diff
@@ -56,8 +56,31 @@ void my_mergesort(int left, int right) {
+		my_mergesort(argument->left, argument->right);
+	} else {
+		struct argument *args1 = buildArgs(argument->left, mid, argument->level-1);
+		struct argument *args2 = buildArgs(mid+1, argument->right, argument->level-1);
```
```suggestion
		struct argument* args1 = buildArgs(argument->left, mid, argument->level-1);
		struct argument* args2 = buildArgs(mid+1, argument->right, argument->level-1);
```
Despite the skeleton code insisting on using pointer keyword next to the name. PLEASE pointer keyword next to the type!

Izak8 commented on mergesort.c:
```diff
@@ -56,8 +56,31 @@ void my_mergesort(int left, int right) {
+		pthread_t *p_mergesort2;
+
+		pthread_create(p_mergesort1, NULL, parallel_mergesort, argument);
+		pthread_create(p_mergesort2, NULL, parallel_mergesort, argument);
```
```suggestion
		pthread_create(p_mergesort1, NULL, parallel_mergesort, args1);
		pthread_create(p_mergesort2, NULL, parallel_mergesort, args2);
```

Matthew-ew committed:
fix: add second base case to `parallel_mergesort`

This may not be valid, but it seems correct. If there are enough levels to perform mergesort entirely in parallel then this is the standard mergesort base case.

Co-authored-by: Izak Baldacchino <83893457+Izak8@users.noreply.github.com>

Matthew-ew committed:
fix: pointer keyword next to type like a normal person!

Co-authored-by: Izak Baldacchino <83893457+Izak8@users.noreply.github.com>

Matthew-ew committed:
fix: replaced incorrect argument struct passed to `pthread_create`

Co-authored-by: Izak Baldacchino <83893457+Izak8@users.noreply.github.com>

Izak8 reviewed (approved):
I think this is correct. We will need to do integration testing of course.
